### ISSUE TITLE
ServiceBus response taking over 20 seconds

### ISSUE THREAD

**Issue by jeffGillings** [Xabaril/AspNetCore.Diagnostics.HealthChecks#300]
I'm using HealthChecks AzureServiceBus v2.2.5 to make the calls below

```
services.AddHealthChecks()
                .AddCheck("self", () => HealthCheckResult.Healthy())
                .AddAzureKeyVault(x => x.UseKeyVaultUrl(configuration["KeyVault:Url"]), "idt-kv")
                .AddAzureServiceBusQueue(configuration["SendEmailQueue:ConnectionString"], emailQueueName, emailQueueName)
                .AddAzureServiceBusTopic(configuration["UserAccountDeletedTopic:ConnectionString"], userAccountDeletedTopicName, userAccountDeletedTopicName)
                .AddAzureServiceBusTopic(configuration["UserRegisteredTopic:ConnectionString"], userRegisteredTopicName, userRegisteredTopicName)
```

and notice that I get intermittently failures (between 2-3 a week and sometimes once a month) from the AddAzureServiceBusQueue and AddAzureServiceBusTopic calls (response times exceeding 20 seconds) which I'm only seeing for healthcheck requests.
Any suggestions as to why this is happening?

**Comment by CarlosLanderas** [comment]
Since version 3.0 all healthchecks have a timeout parameter. This is something that was missing in version 2.2 in the Microsoft API and was really important when a service is down, because it takes several seconds to respond.

**Comment by jeffGillings** [comment]
I don't believe that the services are down as it's only the heath check calls that have this issue, application calls to these services don't show this issue and subsequent health check calls (about a minute later) respond in tens of milliseconds.

**Comment by CarlosLanderas** [comment]
Ok. Let me take a look 👍

**Comment by unaizorrilla** [comment]
Hi @jeffGillings 

I'm trying to check Azure Service bus and all is working fine for me! I add the same configuration as you ( one queue and one topic ) .

The health check times are ok after some working hours

![imagen](https://user-images.githubusercontent.com/1630008/66763224-c5621780-eea7-11e9-812f-c7901313b9d7.png)

Probably I need more information about your configuration in order to continue to review this issue:

 - Is healtcheck 2.2 or 3.0?
 - Queue is partitioned? with sessions? with duplicate detection?
 - The same for topic, is partitioned? with duplicate detection?
 - How many message you have on queue/topic?

I try to explain the code of the healthcheck, to check if you can discover any issue.

 The healtcheck create a Queue/TopicClient and reuse this for all the health tests

 ```
                if (!_queueClientConnections.TryGetValue(connectionKey, out var queueClient))
                {
                    queueClient = new QueueClient(_connectionString, _queueName,ReceiveMode.PeekLock, RetryPolicy.NoRetry);

                    if (!_queueClientConnections.TryAdd(connectionKey, queueClient))
                    {
                        return new HealthCheckResult(context.Registration.FailureStatus, description: "New QueueClient connection can't be added into dictionary.");
                    }
                }
```

After create the client, healthcheck schedule a new messge and if the operation success this scheduling is cancelled.

 ```
                var message = new Message(Encoding.UTF8.GetBytes(TEST_MESSAGE));
                _configuringMessage?.Invoke(message);
                
                var scheduledMessageId = await queueClient.ScheduleMessageAsync(message, 
                    new DateTimeOffset(DateTime.UtcNow).AddHours(2));

                await queueClient.CancelScheduledMessageAsync(scheduledMessageId);
```

If your healtcheck is delaying X seconds is becaused scheduling message is delaying X seconds, can you try some simple code to check if you can repro this?

**Comment by unaizorrilla** [comment]
Enjoy guy!! no worries

**Comment by tzkhan** [comment]
Hey @unaizorrilla I'm working on the same codebase alongside @jeffGillings and can help provide some of this information.

- Is healtcheck 2.2 or 3.0?
  - 2.2.5
- Queue is partitioned? with sessions? with duplicate detection?
  - Message TTL - 10675199 days
  - Lock duration - 1 minute
  - Duplicate detection history - 10 minutes
  - Max size - 5GB
  - Current size - 0 KB
  - Partioned - false
- The same for topic, is partitioned? with duplicate detection?
  - Message TTL - 10675199 days
  - Duplicate detection history - 10 minutes
  - Max size - 5GB
  - Current size - 15405.42578125 KB
  - Partioned - false
  - There's a single subscription on the topic with the following settings:
    - Message TTL - 14 days
    - Lock duration - 30 seconds
    - Max delivery count - 10
- How many message you have on queue/topic?
  - on the queue, it's usually 0
  - on one of the recently failing topic subscription, there're currently 18206 messages

Basically, the healthcheck is fine most of the time but fails very intermittently sometimes. For instance, the topic/sub configured with the settings mentioned above has returned an unhealthy status at the following times today:
- 7:07:54.023
- 7:07:43.501 
- 6:07:43.596 
- 6:07:52.735 
- 6:37:42.993 
- 6:37:43.466 

The logs don't seem to help much as this is what gets logged:

`{ Id: 103, Name: "HealthCheckEnd" } Health check user-profile-upserted-topic completed after 19938.1855ms with status Degraded and 'null'`

Is there any way to turn on logs and increase their verbosity for builtin healthchecks? I think we've tried setting the log level for `Microsoft.Extensions.Diagnostics.HealthChecks` to `Information` but that didn't have much of a difference.

**Comment by sungam3r** [comment]
It is possible that this is a client library error. Much time has passed since. Is this problem still relevant?

**Comment by sungam3r** [comment]
Closed due to lack of feedback.
